src/Processors/Formats/Impl/PrettyBlockOutputFormat.cpp

```diff
@@ -509,6 +488,53 @@ void PrettyBlockOutputFormat::writeValueWithPadding(
     }
 }
 
+void PrettyBlockOutputFormat::writeHeaderWithPadding(
+    const Widths & max_widths, const Widths & name_widths, const char * separator, size_t col_num, const ColumnWithTypeAndName & col)
+{
+    String visible_col_name = col.name;
+    UInt64 name_width = name_widths[col_num];
+    if (name_width > max_widths[col_num])
+    {
+        const char * ellipsis = format_settings.pretty.charset == FormatSettings::Pretty::Charset::UTF8 ? "⋯" : "~";
+        const size_t front_bytes = UTF8::computeBytesBeforeWidth(
+            reinterpret_cast<const UInt8 *>(col.name.data()),
+            col.name.size(),
+            0,
+            static_cast<size_t>(std::ceil((max_widths[col_num] - 1) / 2.0)));
+
+        const size_t back_width = static_cast<size_t>(std::floor((max_widths[col_num] - 1) / 2.0));
+        size_t cur_back_width = 0;
+        size_t pos = col.name.size()-1;
+        while (cur_back_width < back_width)
+        {
+            cur_back_width = UTF8::computeWidth(reinterpret_cast<const UInt8 *>(col.name.data()+pos), col.name.size()-pos, 0);
+            --pos;
+        }
+        visible_col_name = col.name.substr(0, front_bytes) + ellipsis + col.name.substr(pos+1);
+        name_width = max_widths[col_num];
```

Comment:
I wonder `name_width` should be the actual width of `visible_col_name`? if `cur_back_width` is large than `back_width` under some boundary conditions, `name_width` may be large than `format_settings.pretty.max_column_name_width`. I made a little change and not sure if it is correct:  
```
size_t cur_back_width = 0;
size_t pos = col.name.size()-1;
size_t last_back_width = cur_back_width;
size_t last_pos = pos;
while (cur_back_width < back_width)
{
    cur_back_width = UTF8::computeWidth(reinterpret_cast<const UInt8 *>(col.name.data()+pos), col.name.size()-pos, 0);
    --pos;

    if (last_back_width != cur_back_width && cur_back_width <= back_width)
    {
        last_back_width = cur_back_width;
        last_pos = pos;
    }
}

visible_col_name = col.name.substr(0, front_bytes) + ellipsis + col.name.substr(last_pos+1);
visible_col_name_width = std::min(UTF8::computeWidth(reinterpret_cast<const UInt8 *>(visible_col_name.data()), visible_col_name.size()), max_widths[col_num]);
```

Reply:
I'd do this as well.